# Networks: structure, evolution & processes
**Internet Analytics - Lab 2**

---

**Group:** K

**Names:**

*  Xavier JeanMonod
*  Adrian Baudat
*  Simon Wicky

---

#### Instructions

*This is a template for part 1 of the lab. Clearly write your answers, comments and interpretations in Markodown cells. Don't forget that you can add $\LaTeX$ equations in these cells. Feel free to add or remove any cell.*

*Please properly comment your code. Code readability will be considered for grading. To avoid long cells of codes in the notebook, you can also embed long python functions and classes in a separate module. Don’t forget to hand in your module if that is the case. In multiple exercises, you are required to come up with your own method to solve various problems. Be creative and clearly motivate and explain your methods. Creativity and clarity will be considered for grading.*

---

## 2.1 Exploration of real networks

#### Exercise 2.1

In [25]:
import networkx as nx
import numpy as npy

In [26]:
Data = open('../data/network1.csv', "r")
next(Data,None)
Graphtype = nx.Graph()
G = nx.parse_edgelist(Data, delimiter=',', create_using=Graphtype,
                      nodetype=int)
print("Number of edges : " + str(len(G.edges)))
print("Number of nodes : " + str(len(G.nodes)))

Number of edges : 18584
Number of nodes : 13033


#### Exercise 2.2:  Node degree distribution

#### Exercise 2.3: Giant component

In [38]:

print("Number of connect components : " + str(nx.number_connected_components(G)))
size = [len(o) for o in nx.connected_components(G)]

print("Histogrm of connected components (size : number of CC)")
for i in range(max(size) + 1):
    if size.count(i) > 0:
        print(str(i) + " : " + str(size.count(i)))
    

Number of connect components : 192
Histogrm of connected components (size : number of CC)
2 : 149
3 : 29
4 : 7
5 : 3
6 : 1
7 : 1
16 : 1
12576 : 1


Looking at the data above, we can see that there is 192 connected components, with one giant components of size 12756

#### Exercise 2.4: Short paths & Small-worlds

In [68]:
giant_cc = [e for e in nx.connected_component_subgraphs(G) if len(e.nodes()) == max(size)][0]
print(nx.average_shortest_path_length(giant_cc))

NodeDataView({0: None, 11932: None, 1: None, 5128: None, 2: None, 7008: None, 3: None, 10476: None, 4: None, 9671: None, 5: None, 1696: None, 7185: None, 6702: None, 6: None, 875: None, 11859: None, 7: None, 4413: None, 8: None, 6753: None, 9: None, 3467: None, 3644: None, 10: None, 5094: None, 11: None, 6658: None, 12: None, 461: None, 13: None, 3961: None, 14: None, 16: None, 1422: None, 17: None, 3110: None, 18: None, 19: None, 4890: None, 20: None, 870: None, 7607: None, 21: None, 2598: None, 22: None, 8094: None, 23: None, 24: None, 25: None, 5126: None, 26: None, 2153: None, 27: None, 28: None, 619: None, 6281: None, 10667: None, 818: None, 4739: None, 6364: None, 8853: None, 7433: None, 11401: None, 2554: None, 12892: None, 5951: None, 29: None, 9770: None, 30: None, 5547: None, 9796: None, 3158: None, 6382: None, 31: None, 2449: None, 32: None, 33: None, 6588: None, 34: None, 3728: None, 35: None, 9900: None, 36: None, 10753: None, 1561: None, 37: None, 1941: None, 38: None, 16

#### Exercise 2.5: Network comparison

#### Exercise 2.6: Network identification